### PR TITLE
Dp 392 Add refresh functionality to KinClient in Kin SDK

### PR DESCRIPTION
**Pull request by drEfrati**
In order for all instances of KinClient to be always up to date regarding the accounts. When getting an account or getting the size of the accounts, etc.. then we make sure that the account list is being up to date.

### CODE REVIEW

File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java [hunk 1]
```diff
@@ -215,7 +230,6 @@ public void deleteAccount(int index) throws DeleteAccountException {
     /**
      * Deletes all accounts.
      */
-    @SuppressWarnings("WeakerAccess")
     public void clearAllAccounts() {
         keyStore.clearAllAccounts();
```
Comment (yosriz) [inline kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java]: don't we need a `loadAccount` here as well?

File: kin-sdk/kin-sdk-lib/src/test/java/kin/sdk/KinClientTest.java [hunk 2]
```diff
@@ -161,6 +158,39 @@ public void getAccount_ExistingMultipleAccount() throws Exception {
         assertThat(account2.getAccountId(), equalTo(expectedAccount2.getPublicAddress()));
     }
 
+    @Test
+    public void createMultipleKinClients_SameAccounts() throws Exception {
```
Comment (yosriz) [inline kin-sdk/kin-sdk-lib/src/test/java/kin/sdk/KinClientTest.java]: 👍

File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java [hunk 3]
```diff
@@ -111,10 +108,26 @@ private void loadAccounts() {
...
         }
         if (accounts != null && !accounts.isEmpty()) {
-            for (KeyPair account : accounts) {
-                kinAccounts.add(createNewKinAccount(account));
+            updateKinAccounts(accounts);
+        }
+    }
+
+    private void updateKinAccounts(List<KeyPair> storageAccounts) {
+        Map<String, KinAccountImpl> accountsMap = new HashMap<>();
```
Comment (yosriz) [inline kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java]: BTW, you should be able to do this just with `Set`, but I can see that we have only `equals` implemented without `hashCode`, so probably that's why you didn't use it.

Comment (drEfrati) [reply kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java]: yes, although we can generate a hashCode with the help of the IDE.
Keep it that way or to change?

File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java (hunk 1 above)
Comment (drEfrati) [reply kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java]: I added loadAccount in a "lazy" way, meaning only when someone acctualy tries to do something, meaning getAccount, getAccountCount and getAccountByPublicAddress.
Do you think we must add it also to clearAllAccounts, deleteAccount and addKeyPair?

**Comment by drEfrati** [comment]
> @drEfrati good work, some comments:
> 
> * I think we need to add `loadAccounts` to `clearAllAccounts`, `deleteAccount` and `addKeyPair` as well, you can do this TDD style, add tests that I assume will fail, and then fix them by adding the call to `loadAccounts`.
> * Add tests for the above scenarios as well for the import account scenario which is important.

See my comment

File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java (hunk 1 above)
Comment (yosriz) [reply kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java]: I think it's a potential bug, if you delete an account on kinClient1, while kinClient2 added an account, the last will win, I would expect one account to be added and one be deleted.

File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java (hunk 3 above)
Comment (yosriz) [reply kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java]: not critical at all, your call

File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java (hunk 1 above)
Comment (drEfrati) [reply kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java]: but at the second one of the kinClient will try to use one of the methods that query the account(getAccount or getAccountCount or getAccountByPublicAddress) then he will get the correct result and the internal list will be updated 
**for example:**
current account in database are("1", "2", "3", 4") then kinClient1 delete account "2" and then kinClient2 add account "5". the database will now look like this ("1", "3", "4", "5").
Although the internal list for each one of them will be different,  at the second that one of them will invoke one of the above methods then the internal list of both of them will be updated.

I added another test for this case.
What is your concern exactly? or you don't like the idea that the internal state is temporary not the same?

Comment (drEfrati) [reply kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java]: @yosriz please review again after my last changes

File: kin-sdk/kin-sdk-lib/src/test/java/kin/sdk/FakeKeyStore.java [hunk 4]
```diff
@@ -24,8 +25,13 @@
     }
 
     @Override
-    public void deleteAccount(int index) {
-        accounts.remove(index);
+    public void deleteAccount(String publicAddress) {
+        for (int i = 0; i < accounts.size(); i++) {
+            if (accounts.get(i).getAccountId().equals(publicAddress)) {
+                accounts.remove(i);
```
Comment (yosriz) [inline kin-sdk/kin-sdk-lib/src/test/java/kin/sdk/FakeKeyStore.java]: removing in the middle of the loop can cause errors

File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java [hunk 5]
```diff
@@ -189,6 +189,7 @@ private KinAccount addKeyPair(KeyPair account) {
     }
 
     /**
+     * @deprecated please muse getAccount(String publicAdderss) will be remove is cersion 2.2.3
```
Comment (yosriz) [inline kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java]: remove this comment for now

File: kin-sdk/kin-sdk-lib/src/test/java/kin/sdk/FakeKeyStore.java (hunk 4 above)
Comment (drEfrati) [reply kin-sdk/kin-sdk-lib/src/test/java/kin/sdk/FakeKeyStore.java]: As far as I know, it is correct if you continue with the loop but in this case immediately after I found the item I remove and stop the loop.
It is the same as to find the index exit the loop and then removing, no?
And it is in the tests only.

File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java (hunk 5 above)
Comment (drEfrati) [reply kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinClient.java]: forgot it...

File: kin-sdk/kin-sdk-lib/src/test/java/kin/sdk/FakeKeyStore.java (hunk 4 above)
Comment (yosriz) [reply kin-sdk/kin-sdk-lib/src/test/java/kin/sdk/FakeKeyStore.java]: all cool, missed the break, need new 👓

Comment (drEfrati) [reply kin-sdk/kin-sdk-lib/src/test/java/kin/sdk/FakeKeyStore.java]: Already update the code, take a look

**Pull request merged**